Sources/OpenCombine/Publishers/ChildManagingInnerBase.swift

```diff
@@ -0,0 +1,327 @@
+                }
+            }
+
+            if newDemand != .none {
+                lock.do { downstreamDemand += newDemand }
+            }
+        }
+
+        // Loop to try again in case somebody put stuff on the queue while we were
+        // sending values with the lock unlocked.
+        processQueue()
+    }
+}
+
+// InnerWithChildrenBase is the `Subscription` for `Downstream`
+extension InnerWithChildrenBase: Subscription {
+    internal func request(_ demand: Subscribers.Demand) {
+        let (drainTheQueue, demandModeChanged) = lock.do { () -> (Bool, Bool) in
+            let demandModeChanged =
+                (demand == .unlimited) != (downstreamDemand == .unlimited)
+            downstreamDemand = demand
+            if processingQueue {
+                return (false, demandModeChanged)
+            }
+            processingQueue = true
+            return (true, demandModeChanged)
+        }
+
+        if demandModeChanged {
+            // TODO: This code is currently asymmetric because Apple only updates the
+            // child demand when going from limited to unlimited. It does not update the
+            // demand when goinf from unlimited to limited (which seems like a bug to me).
+            if demand == .unlimited {
+                // TODO: This code isn't yet thread safe. The correct change is to do this
+                // through the queue just like sending values and finished. Finished is
+                // done through the queue as a bit of a hack. The right design is to have
+                // an enum of actions on the queue. That enum will include (send value,
+                // send finished, set child demand).
+                let newChildDemand = demandForChild()
+                childSubscribers.forEach { $0.request(newChildDemand) }
+            }
+        }
+
+        if drainTheQueue {
+            processQueue()
+        }
+    }
+}
+
+extension InnerWithChildrenBase {
+    /// ChildSubscriber is needed to help implement the backpressure/demand strategy.
+    /// Specifically, a custom subscriber is needed to manage the demand of the child
+    /// subscription:
+    ///  - Send .max(1) request when the subscription is received
+    ///  - Send .max(1) request when downstream subscriber demands more and a previously
+    ///   buffered value from the child was sent. (When the value was buffered, the
+    ///   child's demand reached .none - effectively pausing the child.)
+    internal final class ChildSubscriber: Hashable {
+        internal typealias Input = Downstream.Input
+        internal typealias Failure = Downstream.Failure
+
+        private var _upstreamSubscription: Subscription?
+        private let _receiveValue: (ChildSubscriber, Input) -> Subscribers.Demand
+        private let _receiveCompletion: (ChildSubscriber, Subscribers.Completion<Failure>)
+        -> Void
+        private let _initialDemandForChild: () -> Subscribers.Demand
```

Comment:
Yeah. My recent realization about makes this all moot. I'll be getting rid of the base class since I actually think that there are no other consumers of it.